File: algorithm/1211/모래성.py
dr = (-1, -1, -1, 0, 0, 1, 1, 1)
dc = (-1, 0, 1, -1, 1, -1, 0, 1)

H, W = map(int, input().split())
q = []
arr = [[] for _ in range(H)]

for r in range(H):
    temp = list(input())
    for c in range(W):
        if temp[c] == '.':
            temp[c] = 0
        else:
            temp[c] = int(temp[c])
    arr[r] = temp

life = [arr[r][:] for r in range(H)]

for r in range(H):
    for c in range(W):
        if arr[r][c] == 0:
            continue
        count = 0
        for d in range(8):
            nr = r + dr[d]
            nc = c + dc[d]
            if not (H > nr >= 0 or W > nc >= 0):
                continue
            if arr[nr][nc] == 0:
                count += 1
        life[r][c] -= count
        if life[r][c] <= 0:
            q.append((r, c))

result = 0
while q:
    result += 1
    time = len(q)
    for _ in range(time):
        r, c = q.pop(0)z
        for d in range(8):
            nr = r + dr[d]
            nc = c + dc[d]
            if not (H > nr >= 0 or W > nc >= 0):
                continue
            if life[nr][nc] <= 0:
                continue
            life[nr][nc] -= 1
            if life[nr][nc] <= 0:
                q.append((nr, nc))

print(result)

File: algorithm/c++/algorithm/모래성/모래성.cpp
#include <iostream>
#include <queue>

using namespace std;

int H, W, time, nr, nc, waveCount, result;
int arr[1000][1000];
int dr[] = { -1, -1, -1, 0, 0, 1, 1, 1 };
int dc[] = { -1, 0, 1, -1, 1, -1, 0, 1 };
char chr;

typedef struct point {
	int pr = 0;
	int pc = 0;
} point;

queue<point> q;

int main() {
	scanf("%d %d\n", &H, &W);
	
	for (int r = 0; r < H; ++r) {
		for (int c = 0; c < W; ++c) {
			scanf("%c\n", &chr);
			if (chr == '.') arr[r][c] = 0;
			else arr[r][c] = chr - '0';
		}
	}

	for (int r = 0; r < H; ++r) {
		for (int c = 0; c < W; ++c) {
			waveCount = 0;
			for (int i = 0; i < 8; ++i) {
				nr = dr[i] + r;
				nc = dc[i] + c;
				if (nr < 0 || nc < 0 || nr >= H || nc >= W) continue;
				if (arr[nr][nc] == 0) waveCount++;
			}
			point a = new point;
			if (waveCount >= arr[r][c]) q.push(a);
		}
	}

	while (!q.empty()) {
		time = q.size();
		for (int i = 0; i < time; ++i) {
			q.pop();
		}
	}

	return 0;
}